server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java

```diff
@@ -278,6 +279,14 @@ protected static List<QueryBuilder> readQueries(StreamInput in) throws IOExcepti
 
     @Override
     public final QueryBuilder rewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        QueryRewriteInterceptor queryRewriteInterceptor = queryRewriteContext.getQueryRewriteInterceptor();
+        if (queryRewriteInterceptor != null) {
+            var rewritten = queryRewriteInterceptor.rewrite(queryRewriteContext, this);
+            if (rewritten != this) {
+                return rewritten;
```

Comment:
It was suggested that I wrap this `rewritten` in a `InterceptedQueryBuilderWrapper` but I simply could not get this to work as expected using just `QueryBuilder.rewrite`. I think the current refactoring is still cleaner and an improvement, though, so I hope we should be OK here.